library.py

```diff
@@ -0,0 +1,174 @@
+import torch
+import inspect
+import functools
+
+"""
+These are changes that we'll add to PyTorch
+"""
+
+
+class OpDef:
+    @classmethod
+    def build(cls):
+        return register(cls)
+
+
+OPERATOR_REGS = {}
+
+
+def register(opdef):
+    name = opdef.__name__
+
+    frm = inspect.stack()[2]
+    mod = inspect.getmodule(frm[0])
+    ns = mangle_module(mod.__name__)
+
+    check_allowed_attrs(opdef)
+    qualname = f"{ns}::{name}"
+    lib = get_library_allowing_overwrite(ns, name)
+    lib.define(f"{name}{opdef.schema}")
+    op = torch._library.utils.lookup_op(qualname)
+
+    impl_method = getattr(opdef, "impl", None)
+    if impl_method is not None:
+        properties = getattr(impl_method, "__properties__", {})
+        for device in properties["device_types"]:
+            register_backend(lib, name, impl_method, device)
+
+    impl_cpu_method = getattr(opdef, "impl_cpu", None)
+    if impl_cpu_method is not None:
+        register_backend(lib, name, impl_cpu_method, "CPU")
+
+    impl_cuda_method = getattr(opdef, "impl_cuda", None)
+    if impl_cuda_method is not None:
+        register_backend(lib, name, impl_cuda_method, "CUDA")
+
+    properties = getattr(impl_method, "__properties__", {}) if impl_method is not None else {}
+    if getattr(opdef, "abstract", None):
+        torch.library.impl_abstract(qualname, opdef.abstract, lib=lib)
+    elif impl_method and properties.get('traceable', False):
+        torch.library.impl_abstract(qualname, impl_method, lib=lib)
+
+    register_autograd(lib, name, opdef, op)
+
+    return op
+
+
+def check_allowed_attrs(op):
+    return
+    attrs = set(dir(op)) - set(dir(object))
+    allowed_attrs = {
+        "impl_cpu",
+        "impl_cuda",
+        "setup_backward",
+        "backward",
+    }
+    if attrs.issubset(allowed_attrs):
+        return
+    raise RuntimeError("not subset")
+
+
+def get_library_allowing_overwrite(ns, name):
+    qualname = f"{ns}::{name}"
+
+    if qualname in OPERATOR_REGS:
+        OPERATOR_REGS[qualname]._destroy()
+        del OPERATOR_REGS[qualname]
+
+    lib = torch.library.Library(ns, "FRAGMENT")
+    OPERATOR_REGS[qualname] = lib
+    return lib
+
+
+def traceable(f):
+    @functools.wraps(f)
+    def wrapper(*args, **kwargs):
```

Comment:
Why a wrapper and not directly on f? Static method shenanigans? :p

Reply:
yes